### PR TITLE
Optimize layer normalization

### PR DESCRIPTION
**Pull request by soskek**
The current implementation of layer normalization is slow (as [this](https://github.com/chainer/chainer/pull/1950#issuecomment-265613756) mentioned). I implemented forward and backward computation explicitly as a new function.

### CODE REVIEW

**Comment by soskek** [comment]
Failures by Appveyor look not due to this PR.

File: chainer/functions/normalization/layer_normalization.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        g_scaled_x = gy
+
+        g_gamma = xp.sum(g_scaled_x * self.x_hat, axis=0)
+        g_x_hat = g_scaled_x * gamma[None, ]
+
+        g_inv_std = xp.sum(g_x_hat * self.x_mu, axis=1, keepdims=True)
+        g_x_mu_1 = g_x_hat * self.inv_std
+
+        g_std = g_inv_std * (- 1. / self.var)
+        # = g_inv_std * (- 1. / (self.std ** 2))
```
Comment (mitmul) [inline chainer/functions/normalization/layer_normalization.py]: Could you remove this comment line?

File: chainer/functions/normalization/layer_normalization.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+        g_var = g_std * 0.5 * self.inv_std
+        # = g_std * 0.5 * 1. / xp.sqrt(self.var + self.eps)
```
Comment (mitmul) [inline chainer/functions/normalization/layer_normalization.py]: Ditto.

File: chainer/functions/normalization/layer_normalization.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        n_units = x.shape[1]
+        g_squ_x_mu = _broadcast_to(xp, g_var * 1. / n_units, x.shape)
+        g_x_mu_2 = g_squ_x_mu * 2 * self.x_mu
+
+        g_x_1 = g_x_mu_1 + g_x_mu_2
+        g_mu = xp.sum(g_x_1, axis=1, keepdims=True) * (- 1.)
+        # = xp.sum(g_x_mu_1 + g_x_mu_2, axis=1, keepdims=True) * (- 1.)
```
Comment (mitmul) [inline chainer/functions/normalization/layer_normalization.py]: Ditto.

**Comment by mitmul** [comment]
I passed all CI tests. I'll merge this after the above comment lines are removed.

File: chainer/functions/normalization/layer_normalization.py (hunk 1 above)
Comment (soskek) [inline chainer/functions/normalization/layer_normalization.py]: Sure. Thank you for review!!

**Comment by mitmul** [comment]
LGTM

**Pull request merged**